added testcases for LazyVideos and the corresponding .ccz as well

kbo001 opened the pull request:
## Summary
Automated the testcases for VideosWithValidReferences and VideosWithNoReference in the Lazy Videos Tab of the Mobile Test Plan:
https://docs.google.com/spreadsheets/d/1X9YAf_g1Cq9vZZdCkcBLnRSFrhZuk_vEyCSM3sD5QvM/edit#gid=1556040007
Also added it corresponding .ccz file

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Product Description
<!-- For non-invisible changes, describe user-facing effects. Would be good to add screenshots/videos for any major user facing changes -->

## Safety Assurance

- [ ] If the PR is high risk, "High Risk" label is set
- [ ] I have confidence that this PR will not introduce a regression for the reasons below
- [ ] Do we need to enhance manual QA test coverage ? If yes, "QA Note" label is set correctly

### Automated test coverage
This testcase covers the below scenarios:
1. opens the form Video with valid reference , checks if the video is already downloaded while installing the app (if not already present then downloads it), plays the video, moves to next page and verifies that the video player is present and then submits the form.
2. opens the form Video with no reference , checks the video is not downloaded while installing the app, tries to download the video, gets error "Media not found in the application", moves to next page and verifies that the video player is missing and then submits the form.

### Safety story
Executed in local AVD, ran successfully.

dimagimon commented:
<!--
  0 Errors
  1 Warning: This PR does not contain any J...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR does not contain any JIRA issue keys in the PR title or commit messages (e.g. KEY-123)</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

shubham1g5 reviewed (changes requested):
@kbo001 I was reading the steps on manual test plan for this and it seems to state that we need to repeat the steps for all questions in the form (steps 3.6 and 5.3 in test plan).  In comparison this test only seem to test the lazy download for the first question in the form.

kbo001 commented on app/instrumentation-tests/src/org/commcare/androidTests/LazyVideosTests.kt:
```diff
@@ -0,0 +1,109 @@
+                )
+            ).isPresent();
+        } else {
+            InstrumentationUtility.stubIntentWithAction(Intent.ACTION_VIEW)
+            onView(withId(R.id.video_button)).perform(ViewActions.click())
+            InstrumentationUtility.nextPage()
```
yes. Removing the else statement

kbo001 commented on app/instrumentation-tests/src/org/commcare/androidTests/LazyVideosTests.kt:
```diff
@@ -0,0 +1,109 @@
+            )
+        ).isPresent()
+        InstrumentationUtility.stubIntentWithAction(Intent.ACTION_SEND)
+        onView(withId(R.id.video_button)).perform(ViewActions.click())
```
The button name is same for both valid and not valid scenarios. However the image in the button differs.

shubham1g5 replied:
got it, thanks!

shubham1g5 commented on app/instrumentation-tests/src/org/commcare/androidTests/LazyVideosTests.kt:
```diff
@@ -0,0 +1,109 @@
+                R.drawable.update_download_icon
+            )
+        ).isPresent()
+        InstrumentationUtility.stubIntentWithAction(Intent.ACTION_SEND)
```
I don't think we need to do this as `Intent.ACTION_SEND` never gets fired during this workflow.

shubham1g5 replied:
@kbo001 FYI

kbo001 replied:
removed

damagatchi commented:
Can one of the admins verify this patch?

shubham1g5 commented:
@damagatchi retest this please

shubham1g5 commented:
@damagatchi retest this please

shubham1g5 commented:
@damagatchi retest this please

shubham1g5 commented:
@damagatchi retest this please

shubham1g5 commented:
@kbo001 Are the changes here complete now ?
@damagatchi retest this please

kbo001 commented:
@shubham1g5 I did not get the time yesterday, working on it today.

kbo001 commented:
@shubham1g5 I changed the ccz file and updated the test steps according to the testplan.

shubham1g5 commented on app/instrumentation-tests/src/org/commcare/androidTests/LazyVideosTests.kt:
```diff
@@ -0,0 +1,152 @@
+        InstrumentationUtility.login("test1", "123")
+        InstrumentationUtility.openForm(1, 0)
+        InstrumentationUtility.waitForView(withId(R.id.video_button))
+        if ((onView(withTagValue(CoreMatchers.`is`(R.drawable.update_download_icon))).isPresent())
```
this should always be present now right ?

kbo001 replied:
yes.

shubham1g5 replied:
we should not put this in `if` condition then

kbo001 replied:
But if we keep it, then even if the .ccz is updated then also this will work. Thats what I thought of. If you think it should be removed, I can do that.

shubham1g5 replied:
We should remove it as we don't want to paas this test unless it goes through the download workflow.

kbo001 replied:
removed

shubham1g5 commented:
`lazy_videos_test.ccz` can be removed now ?

kbo001 commented on app/instrumentation-tests/src/org/commcare/androidTests/LazyVideosTests.kt:
```diff
@@ -0,0 +1,153 @@
+
+    @Before
+    fun setup() {
+        if (CommCareApplication.instance().currentApp == null) {
```
if `installApp()` is used then the app is not uninstalled everytime and the videos remains in the app, hence failing the tests.

kbo001 commented on app/instrumentation-tests/src/org/commcare/androidTests/LazyVideosTests.kt:
```diff
@@ -0,0 +1,153 @@
+        InstrumentationUtility.waitForView(withSubstring("Download in Progress"))
+        InstrumentationUtility.waitForView(withSubstring("Download complete"))
+        InstrumentationUtility.waitForView(withTagValue(CoreMatchers.`is`(android.R.drawable.ic_media_play)))
+//        }
```
removed

shubham1g5 reviewed (approved):
Approved given the test pass on BS

shubham1g5 commented:
@kbo001 Seems like DateWidgetTests are failing again repeatedly blocking pass on this test. Would you be able to take a look. 
@damagatchi retest this please

kbo001 commented:
@shubham1g5 the dateWidgetTests seems to be passing but the lazy videos is failing as in the third question the video was still not available. I have added some wait time during the video download and also added the video download code for the remaining questions. Can you execute it once on BS as its already passing at my end.

kbo001 commented:
I see that the test is passing now

shubham1g5 commented on app/instrumentation-tests/src/org/commcare/androidTests/LazyVideosTests.kt:
```diff
@@ -93,14 +93,17 @@ class LazyVideosTests : BaseTest() {
         onView(withTagValue(CoreMatchers.`is`(R.drawable.update_download_icon))).perform(click())
-        onView(withId(R.id.video_button)).perform(click())
+        InstrumentationUtility.sleep(1)
+        InstrumentationUtility.waitForView(withSubstring("Download started"))
```
this should be immediate after click, wait before and after this should not be required.

kbo001 replied:
This video button will be available only when the download is completed. L104 is doing that task here. Once the download is complete we click on the video button and verify the intent. We don't need it here as we will be clicking on `R.drawable.update_download_icon` and start the download.

shubham1g5 replied:
I am talking about `InstrumentationUtility.sleep()`  statements, I don't think they should be required before and after this statement since we are already using `waitForView` here.

shubham1g5 commented on app/instrumentation-tests/src/org/commcare/androidTests/LazyVideosTests.kt:
```diff
@@ -114,6 +117,15 @@ class LazyVideosTests : BaseTest() {
                 R.drawable.icon_info_outline_lightcool
             )
         ).perform(click())
+        if(onView(withTagValue(CoreMatchers.`is`(R.drawable.update_download_icon))).isPresent()){
```
why do we have a `if` here ?

kbo001 replied:
Same reason as the one before. Sometimes the videos for the other questions may not get downloaded properly.

kbo001 commented on app/instrumentation-tests/src/org/commcare/androidTests/LazyVideosTests.kt:
```diff
@@ -126,6 +138,15 @@ class LazyVideosTests : BaseTest() {
                 R.drawable.icon_info_outline_lightcool
             )
         ).perform(click())
+        if(onView(withTagValue(CoreMatchers.`is`(R.drawable.update_download_icon))).isPresent()){
```
The test was failing earlier because of this, instead of the play icon there was the download icon, maybe the video for this question was not downloaded properly. This time when the test is executed the download icon was not present as earlier. This might happen again so I kept this condition for both the last questions.

shubham1g5 replied:
If we put these in a `if` we are bypassing the normal test steps to download media and a pass here won't guarantee that the underlying download workflow works properly risking a bug into the mobile release.

kbo001 commented:
@shubham1g5 I have commented the sleep and if statements, can you verify if the test is passing in BS?
